docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,295 @@
+
+```go
+func (rs *Store) CacheMultiStore() types.CacheMultiStore {
+	stores := make(map[types.StoreKey]types.CacheWrapper)
+	for k, v := range rs.stores {
+		stores[k] = v
+	}
+	var cacheListeners map[types.StoreKey][]types.WriteListener
+	if rs.cacheListening {
+		cacheListeners = rs.listeners
+	}
+	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, cacheListeners)
+}
+```
+
+### Exposing the data 
+We will introduce and document mechanisms for exposing data from the above listeners to external consumers.
+
+#### Writing to file
+We will document and provide examples of how to configure a listener to write out to a file.
+No new type implementation will be needed, an `os.File` can be used as the underlying `io.Writer` for a `GobWriteListener`.
+
+Writing to a file is the simplest approach for streaming the data out to consumers.
+This approach also provide the advantages of being persistent and durable, and the files can be read directly
+or an auxiliary streaming services can tail the files and serve the data remotely.
+
+Without pruning the file size can grow indefinitely, this will need to be managed by
+the developer in an application or even module-specific manner.
+
+#### Writing to gRPC stream
+We will implement and document an `io.Writer` type for exposing our listeners over a gRPC server stream.
+
+Writing to a gRPC stream gRPC will allow us to expose the data over the standard gRPC interface.
+This interface can be exposed directly to consumers or we can implement a message queue or secondary streaming service on top.
+Using gRPC will provide us with all of the regular advantages of gRPC and protobuf: versioning guarantees, client side code generation, and interoperability with the many gRPC plugins and auxillary services.
+
+Proceeding through a gRPC intermediate will provide additional overhead, in most cases this is not expected to be rate limiting but in
+instances where it is the developer can implement a more performant streaming mechanism for state listening.
+
+### Configuration
+We will provide detailed documentation on how to configure the state listeners and their external streaming services from within an app's `AppCreator`,
+using the provided `AppOptions`. We will add two methods to the `BaseApp` to enable this configuration:
+
+```go
+// SetCommitMultiStoreListeners sets the KVStore listeners for the provided StoreKey
+func (app *BaseApp) SetCommitMultiStoreListeners(key sdk.StoreKey, listeners []storeTypes.WriteListener) {
+	app.cms.SetListeners(key, listeners)
+}
+
+// SetCacheListening turns on or off listening at the cache layer
+func (app *BaseApp) SetCacheListening(listening bool) {
+	app.cms.CacheListening(listening)
+}
+```
+
+As a demonstration, we will implement the state watching features as part of SimApp.
+For example, the below is a very rudimentary integration of the state listening features into the SimApp `AppCreator` function:
+
+
```

Comment:
We need a brief section on what the actual (TOML) configuration options are.

e.g.

```toml
[store]
  listening = true|false
  # ...
```

Reply:
We need to also probably support a listeners section:

```toml
[store]
  listeners = [
    "file",
    "grpc"
  ]
```

And we may not even need `listening` -- we can just assume enabled|disabled based off of `len(listeners) > 0`.